# Run Any Kind of OLS Regression (ANOVA, GLM, Logit, etc.)

### Authors: Calvin Howard.

#### Last updated: May 5, 2024

Use this to run/test a statistical model (e.g., regression or T-tests) on a spreadsheet containing covariates and brain image (nii/gii) paths. 

Notes:
- For this to work, it must be installed onto wherever you want to run it. You must run:
```
> git clone https://github.com/Calvinwhow/Research.git
> cd into wherever you installed it. 
> pip install -e .
```
- To best use this notebook, you should be familar with GLM design and Contrast Matrix design. See this webpage to get started:
[FSL's GLM page](https://fsl.fmrib.ox.ac.uk/fsl/fslwiki/GLM)

# 00 - Import CSV with All Data
**The CSV is expected to be in this format**
- ID and absolute paths to niftis are critical
```
+-----+----------------------------+--------------+--------------+--------------+
| ID  | Nifti_File_Path            | Covariate_1  | Covariate_2  | Covariate_3  |
+-----+----------------------------+--------------+--------------+--------------+
| 1   | /path/to/file1.nii.gz      | 0.5          | 1.2          | 3.4          |
| 2   | /path/to/file2.nii.gz      | 0.7          | 1.4          | 3.1          |
| 3   | /path/to/file3.nii.gz      | 0.6          | 1.5          | 3.5          |
| 4   | /path/to/file4.nii.gz      | 0.9          | 1.1          | 3.2          |
| ... | ...                        | ...          | ...          | ...          |
+-----+----------------------------+--------------+--------------+--------------+
```

Prep Output Direction

In [1]:
# Specify where you want to save your results to
out_dir = '/Users/cu135/Partners HealthCare Dropbox/Calvin Howard/studies/atrophy_seeds_2023/Figures/supplement_validation_diagnostic/voxelwise'

Import Data

In [2]:
# Specify the path to your CSV file containing NIFTI paths
input_csv_path = '/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_Atrophy/metadata/cleaned_information/discovery.csv'
sheet = None

In [3]:
from calvin_utils.permutation_analysis_utils.statsmodels_palm import CalvinStatsmodelsPalm
# Instantiate the PalmPrepararation class
cal_palm = CalvinStatsmodelsPalm(input_csv_path=input_csv_path, output_dir=out_dir, sheet=sheet)
# Call the process_nifti_paths method
data_df = cal_palm.read_and_display_data()
data_df

,subid,CSF_Cerebellum,CSF_Subcortex,CSF_MTL,CSF_Occipital,CSF_Frontal,CSF_Parietal,CSF_Temporal,GM_Cerebellum,GM_Subcortex,...,CTh_Subcortex,Q4,TOTAL11,TOTALMOD,DX_BASELINE,DX_M12,composite_atrophy_path,csf_atrophy_path,gm_atrophy_path,wm_atrophy_path
0,002_S_0295,-26319.38152,373.297842,-746.688684,-24081.964640,-32607.59211,-18245.868460,-32777.02389,-23248.955500,1038.504856,...,0.0,1.0,3.00,4.00,Normal,Normal,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...
1,002_S_0413,-13670.88871,5009.504947,-449.838607,-8185.340726,-13903.97717,-6191.852471,-13476.49339,-4562.384254,6161.244622,...,0.0,1.0,3.33,4.33,Normal,Normal,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...
2,002_S_0559,-38702.13792,152.392985,-754.706246,-25738.506400,-34449.36704,-33934.639930,-32128.02182,-7651.536038,-611.057402,...,0.0,2.0,6.00,8.00,Normal,Normal,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...
3,002_S_0619,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...
4,002_S_0685,-31092.93296,-22643.911490,-1014.786011,-32806.816640,-18280.08397,-48808.714580,-31102.25568,-7164.119248,-72582.823910,...,0.0,3.0,3.67,6.67,Normal,Normal,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1300,941_S_4365,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...
1301,941_S_4376,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...
1302,941_S_4377,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...
1303,941_S_4420,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,NaN,NaN,NaN,NaN,NaN,NaN,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...



# 01 - Preprocess Your Data

**Handle NANs**
- Set drop_nans=True is you would like to remove NaNs from data
- Provide a column name or a list of column names to remove NaNs from

In [4]:
data_df.columns

Index(['subid', 'CSF_Cerebellum', 'CSF_Subcortex', 'CSF_MTL', 'CSF_Occipital',
       'CSF_Frontal', 'CSF_Parietal', 'CSF_Temporal', 'GM_Cerebellum',
       'GM_Subcortex', 'GM_MTL', 'GM_Occipital', 'GM_Frontal', 'GM_Parietal',
       'GM_Temporal', 'WM_Cerebellum', 'WM_Subcortex', 'WM_MTL',
       'WM_Occipital', 'WM_Frontal', 'WM_Parietal', 'WM_Temporal',
       'Visual_Frontal', 'Visual_Parietal', 'Visual_Occipital',
       'Visual_Temporal', 'Visual_Cerebellum', 'Visual_Subcortex',
       'Visual_MTL', 'Diagnosis', 'Sex', 'Age', 'Cohort', 'CTh_Cerebellum',
       'CTh_MTL', 'CTh_Occipital', 'CTh_Frontal', 'CTh_Parietal',
       'CTh_Temporal', 'CTh_Subcortex', 'Q4', 'TOTAL11', 'TOTALMOD',
       'DX_BASELINE', 'DX_M12', 'composite_atrophy_path', 'csf_atrophy_path',
       'gm_atrophy_path', 'wm_atrophy_path'],
      dtype='object')

In [5]:
drop_list = ['Visual_Occipital', 'DX_BASELINE', 'csf_atrophy_path']

In [6]:
data_df = cal_palm.drop_nans_from_columns(columns_to_drop_from=drop_list)
data_df

,subid,CSF_Cerebellum,CSF_Subcortex,CSF_MTL,CSF_Occipital,CSF_Frontal,CSF_Parietal,CSF_Temporal,GM_Cerebellum,GM_Subcortex,...,CTh_Subcortex,Q4,TOTAL11,TOTALMOD,DX_BASELINE,DX_M12,composite_atrophy_path,csf_atrophy_path,gm_atrophy_path,wm_atrophy_path
0,002_S_0295,-26319.38152,373.297842,-746.688684,-24081.964640,-32607.592110,-18245.868460,-32777.02389,-23248.955500,1038.504856,...,0.0,1.0,3.00,4.00,Normal,Normal,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...
1,002_S_0413,-13670.88871,5009.504947,-449.838607,-8185.340726,-13903.977170,-6191.852471,-13476.49339,-4562.384254,6161.244622,...,0.0,1.0,3.33,4.33,Normal,Normal,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...
2,002_S_0559,-38702.13792,152.392985,-754.706246,-25738.506400,-34449.367040,-33934.639930,-32128.02182,-7651.536038,-611.057402,...,0.0,2.0,6.00,8.00,Normal,Normal,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...
4,002_S_0685,-31092.93296,-22643.911490,-1014.786011,-32806.816640,-18280.083970,-48808.714580,-31102.25568,-7164.119248,-72582.823910,...,0.0,3.0,3.67,6.67,Normal,Normal,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...
5,002_S_0729,-14931.99589,9351.074758,-392.045941,-16190.164180,6146.821509,-3848.640986,-14880.99255,7723.777233,25143.684940,...,0.0,8.0,6.67,14.67,MCI,Alzheimer,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1243,141_S_0790,-18274.00017,3282.586207,-299.272569,-16010.103010,-30354.498030,-23638.487500,-22186.27001,17343.782780,12695.525030,...,0.0,6.0,9.33,15.33,Alzheimer,Alzheimer,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...
1246,141_S_0852,-11199.97032,4181.203209,-1557.320283,-20065.213510,-24016.923460,-26920.294260,-34367.55586,9397.333523,35521.542170,...,0.0,10.0,14.00,24.00,Alzheimer,Alzheimer,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...
1255,141_S_1137,-23089.70453,-10745.059490,-1129.400838,-41824.703730,-36131.115530,-33677.468890,-43382.10995,-6729.084250,616.704174,...,0.0,10.0,26.67,36.67,Alzheimer,Alzheimer,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...
1256,141_S_1152,-13628.75801,-6510.956805,-679.958590,-20290.388240,-4794.264065,-13936.061260,-22084.88688,23967.575890,-40505.874430,...,0.0,9.0,17.00,27.00,Alzheimer,Alzheimer,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...


**Drop Row Based on Value of Column**

Define the column, condition, and value for dropping rows
- column = 'your_column_name'
- condition = 'above'  # Options: 'equal', 'above', 'below'

In [7]:
data_df.columns

Index(['subid', 'CSF_Cerebellum', 'CSF_Subcortex', 'CSF_MTL', 'CSF_Occipital',
       'CSF_Frontal', 'CSF_Parietal', 'CSF_Temporal', 'GM_Cerebellum',
       'GM_Subcortex', 'GM_MTL', 'GM_Occipital', 'GM_Frontal', 'GM_Parietal',
       'GM_Temporal', 'WM_Cerebellum', 'WM_Subcortex', 'WM_MTL',
       'WM_Occipital', 'WM_Frontal', 'WM_Parietal', 'WM_Temporal',
       'Visual_Frontal', 'Visual_Parietal', 'Visual_Occipital',
       'Visual_Temporal', 'Visual_Cerebellum', 'Visual_Subcortex',
       'Visual_MTL', 'Diagnosis', 'Sex', 'Age', 'Cohort', 'CTh_Cerebellum',
       'CTh_MTL', 'CTh_Occipital', 'CTh_Frontal', 'CTh_Parietal',
       'CTh_Temporal', 'CTh_Subcortex', 'Q4', 'TOTAL11', 'TOTALMOD',
       'DX_BASELINE', 'DX_M12', 'composite_atrophy_path', 'csf_atrophy_path',
       'gm_atrophy_path', 'wm_atrophy_path'],
      dtype='object')

Set the parameters for dropping rows

In [8]:
column = 'target'  # The column you'd like to evaluate
condition = 'not'  # The condition to check ('equal', 'above', 'below', 'not')
value = 1 # The value to drop if found

In [9]:
# data_df, other_df = cal_palm.drop_rows_based_on_value(column, condition, value)
display(data_df)

,subid,CSF_Cerebellum,CSF_Subcortex,CSF_MTL,CSF_Occipital,CSF_Frontal,CSF_Parietal,CSF_Temporal,GM_Cerebellum,GM_Subcortex,...,CTh_Subcortex,Q4,TOTAL11,TOTALMOD,DX_BASELINE,DX_M12,composite_atrophy_path,csf_atrophy_path,gm_atrophy_path,wm_atrophy_path
0,002_S_0295,-26319.38152,373.297842,-746.688684,-24081.964640,-32607.592110,-18245.868460,-32777.02389,-23248.955500,1038.504856,...,0.0,1.0,3.00,4.00,Normal,Normal,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...
1,002_S_0413,-13670.88871,5009.504947,-449.838607,-8185.340726,-13903.977170,-6191.852471,-13476.49339,-4562.384254,6161.244622,...,0.0,1.0,3.33,4.33,Normal,Normal,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...
2,002_S_0559,-38702.13792,152.392985,-754.706246,-25738.506400,-34449.367040,-33934.639930,-32128.02182,-7651.536038,-611.057402,...,0.0,2.0,6.00,8.00,Normal,Normal,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...
4,002_S_0685,-31092.93296,-22643.911490,-1014.786011,-32806.816640,-18280.083970,-48808.714580,-31102.25568,-7164.119248,-72582.823910,...,0.0,3.0,3.67,6.67,Normal,Normal,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...
5,002_S_0729,-14931.99589,9351.074758,-392.045941,-16190.164180,6146.821509,-3848.640986,-14880.99255,7723.777233,25143.684940,...,0.0,8.0,6.67,14.67,MCI,Alzheimer,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1243,141_S_0790,-18274.00017,3282.586207,-299.272569,-16010.103010,-30354.498030,-23638.487500,-22186.27001,17343.782780,12695.525030,...,0.0,6.0,9.33,15.33,Alzheimer,Alzheimer,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...
1246,141_S_0852,-11199.97032,4181.203209,-1557.320283,-20065.213510,-24016.923460,-26920.294260,-34367.55586,9397.333523,35521.542170,...,0.0,10.0,14.00,24.00,Alzheimer,Alzheimer,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...
1255,141_S_1137,-23089.70453,-10745.059490,-1129.400838,-41824.703730,-36131.115530,-33677.468890,-43382.10995,-6729.084250,616.704174,...,0.0,10.0,26.67,36.67,Alzheimer,Alzheimer,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...
1256,141_S_1152,-13628.75801,-6510.956805,-679.958590,-20290.388240,-4794.264065,-13936.061260,-22084.88688,23967.575890,-40505.874430,...,0.0,9.0,17.00,27.00,Alzheimer,Alzheimer,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...


Convert Categorical to Ordinal

Misc Changes

In [10]:
import pandas as pd
one_hot = pd.get_dummies(data_df['DX_BASELINE'], prefix='DX_BASELINE', dtype=int)
data_df = data_df.join(one_hot)
data_df

,subid,CSF_Cerebellum,CSF_Subcortex,CSF_MTL,CSF_Occipital,CSF_Frontal,CSF_Parietal,CSF_Temporal,GM_Cerebellum,GM_Subcortex,...,TOTALMOD,DX_BASELINE,DX_M12,composite_atrophy_path,csf_atrophy_path,gm_atrophy_path,wm_atrophy_path,DX_BASELINE_Alzheimer,DX_BASELINE_MCI,DX_BASELINE_Normal
0,002_S_0295,-26319.38152,373.297842,-746.688684,-24081.964640,-32607.592110,-18245.868460,-32777.02389,-23248.955500,1038.504856,...,4.00,Normal,Normal,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,0,0,1
1,002_S_0413,-13670.88871,5009.504947,-449.838607,-8185.340726,-13903.977170,-6191.852471,-13476.49339,-4562.384254,6161.244622,...,4.33,Normal,Normal,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,0,0,1
2,002_S_0559,-38702.13792,152.392985,-754.706246,-25738.506400,-34449.367040,-33934.639930,-32128.02182,-7651.536038,-611.057402,...,8.00,Normal,Normal,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,0,0,1
4,002_S_0685,-31092.93296,-22643.911490,-1014.786011,-32806.816640,-18280.083970,-48808.714580,-31102.25568,-7164.119248,-72582.823910,...,6.67,Normal,Normal,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,0,0,1
5,002_S_0729,-14931.99589,9351.074758,-392.045941,-16190.164180,6146.821509,-3848.640986,-14880.99255,7723.777233,25143.684940,...,14.67,MCI,Alzheimer,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,0,1,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
1243,141_S_0790,-18274.00017,3282.586207,-299.272569,-16010.103010,-30354.498030,-23638.487500,-22186.27001,17343.782780,12695.525030,...,15.33,Alzheimer,Alzheimer,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,1,0,0
1246,141_S_0852,-11199.97032,4181.203209,-1557.320283,-20065.213510,-24016.923460,-26920.294260,-34367.55586,9397.333523,35521.542170,...,24.00,Alzheimer,Alzheimer,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,1,0,0
1255,141_S_1137,-23089.70453,-10745.059490,-1129.400838,-41824.703730,-36131.115530,-33677.468890,-43382.10995,-6729.084250,616.704174,...,36.67,Alzheimer,Alzheimer,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,1,0,0
1256,141_S_1152,-13628.75801,-6510.956805,-679.958590,-20290.388240,-4794.264065,-13936.061260,-22084.88688,23967.575890,-40505.874430,...,27.00,Alzheimer,Alzheimer,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...,1,0,0


# 02 - Define Your Formula

This is the formula relating outcome to predictors, and takes the form:
- y = B0 + B1 + B2 + B3 + . . . BN

It is defined using the columns of your dataframe instead of the variables above:
- 'Apples_Picked ~ hours_worked + owns_apple_picking_machine'

____
**ANOVA**
- Tests differences in means for one categorical variable.
- formula = 'Outcome ~ C(Group1)'

**2-Way ANOVA**
- Tests differences in means for two categorical variables without interaction.
- formula = 'Outcome ~ C(Group1) + C(Group2)'

**2-Way ANOVA with Interaction**
- Tests for interaction effects between two categorical variables.
- formula = 'Outcome ~ C(Group1) * C(Group2)'

**ANCOVA**
- Similar to ANOVA, but includes a covariate to control for its effect.
- formula = 'Outcome ~ C(Group1) + Covariate'

**2-Way ANCOVA**
- Extends ANCOVA with two categorical variables and their interaction, controlling for a covariate.
- formula = 'Outcome ~ C(Group1) * C(Group2) + Covariate'

**Multiple Regression**
- Assesses the impact of multiple predictors on an outcome.
- formula = 'Outcome ~ Predictor1 + Predictor2'

**Simple Linear Regression**
- Assesses the impact of a single predictor on an outcome.
- formula = 'Outcome ~ Predictor'

**MANOVA**
- Assesses multiple dependent variables across groups.
- Note: Not typically set up with a formula in statsmodels. Requires specialized functions.

____
Use the printout below to design your formula. 
- Left of the "~" symbol is the thing to be predicted. 
- Right of the "~" symbol are the predictors. 
- ":" indicates an interaction between two things. 
- "*" indicates and interactions AND it accounts for the simple effects too. 
- "+" indicates that you want to add another predictor. 

In [11]:
data_df.columns

Index(['subid', 'CSF_Cerebellum', 'CSF_Subcortex', 'CSF_MTL', 'CSF_Occipital',
       'CSF_Frontal', 'CSF_Parietal', 'CSF_Temporal', 'GM_Cerebellum',
       'GM_Subcortex', 'GM_MTL', 'GM_Occipital', 'GM_Frontal', 'GM_Parietal',
       'GM_Temporal', 'WM_Cerebellum', 'WM_Subcortex', 'WM_MTL',
       'WM_Occipital', 'WM_Frontal', 'WM_Parietal', 'WM_Temporal',
       'Visual_Frontal', 'Visual_Parietal', 'Visual_Occipital',
       'Visual_Temporal', 'Visual_Cerebellum', 'Visual_Subcortex',
       'Visual_MTL', 'Diagnosis', 'Sex', 'Age', 'Cohort', 'CTh_Cerebellum',
       'CTh_MTL', 'CTh_Occipital', 'CTh_Frontal', 'CTh_Parietal',
       'CTh_Temporal', 'CTh_Subcortex', 'Q4', 'TOTAL11', 'TOTALMOD',
       'DX_BASELINE', 'DX_M12', 'composite_atrophy_path', 'csf_atrophy_path',
       'gm_atrophy_path', 'wm_atrophy_path', 'DX_BASELINE_Alzheimer',
       'DX_BASELINE_MCI', 'DX_BASELINE_Normal'],
      dtype='object')

In [12]:
formula = "composite_atrophy_path ~ DX_BASELINE_Alzheimer + DX_BASELINE_MCI + DX_BASELINE_Normal"

# 02 - Visualize Your Design Matrix

This is the explanatory variable half of your regression formula
_______________________________________________________
Create Design Matrix: Use the create_design_matrix method. You can provide a list of formula variables which correspond to column names in your dataframe.

- voxelwise_variable_list = A list containing the names of each variable that has voxelwise variables. Plainly, the variables that represent niftis. 
- By default, an intercept will be added unless you set intercept=False
- **don't explicitly add the 'intercept' column. I'll do it for you.**

In [13]:
voxelwise_variable_list=['composite_atrophy_path']

If you want to run voxelwise INTERACTIONS, then you should specify the exact terms, exactly as specified in your above formula, here. 
- For example, if Formula is outcome ~ voxelwise_var1 * age + dog_number, then voxelwise_interaction_terms are ['voxelwise_var1 * age]
- Set voxelwise_interaction_terms = [] (left as an empty list) if you have no interactions with voxelwise variables

In [14]:
voxelwise_interaction_terms = []

Make sure ANY voxelwise variables are in formula. 

In [15]:
outcome_df, design_matrix = cal_palm.define_design_matrix(formula, data_df, add_intercept=False,
                                                          voxelwise_variable_list=voxelwise_variable_list, 
                                                          voxelwise_interaction_terms=voxelwise_interaction_terms)
design_matrix

,DX_BASELINE_Alzheimer,DX_BASELINE_MCI,DX_BASELINE_Normal
0,0.0,0.0,1.0
1,0.0,0.0,1.0
2,0.0,0.0,1.0
4,0.0,0.0,1.0
5,0.0,1.0,0.0
...,...,...,...
1243,1.0,0.0,0.0
1246,1.0,0.0,0.0
1255,1.0,0.0,0.0
1256,1.0,0.0,0.0


# 03 - Visualize Your Dependent Variable

I have generated this for you based on the formula you provided

In [16]:
outcome_df

,composite_atrophy_path
0,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...
1,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...
2,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...
4,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...
5,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...
...,...
1243,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...
1246,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...
1255,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...
1256,/Volumes/HowExp/datasets/01a_ADNI_Alzheimer_At...


# 04 - Generate Contrasts

Generate a Contrast Matrix
- This is different from the contrast matrices used in cell-means regressions such as in PALM, but it is much more powerful. 



For more information on contrast matrices, please refer to this: https://cran.r-project.org/web/packages/codingMatrices/vignettes/codingMatrices.pdf

Generally, these drastically effect the results of ANOVA. However, they are mereley a nuisance for a regression.
In essence, they assess if coefficients are significantly different

________________________________________________________________
A coding matrix (a contrast matrix if it sums to zero) is simply a way of defining what coefficients to evaluate and how to evaluate them. 
If a coefficient is set to 1 and everything else is set to zero, we are taking the mean of the coefficient's means and assessing if they significantly
deviate from zero--IE we are checking if it had a significant impact on the ability to predict the depdendent variable.
If a coefficient is set to 1, another is -1, and others are 0, we are assessing how the means of the two coefficients deviate from eachother. 
If several coefficients are 1 and several others are -1, we are assessing how the group-level means of the two coefficients deviate from eachother.
If a group of coefficients are 1, a group is -1, and a group is 0, we are only assessing how the groups +1 and -1 have differing means. 

1: This value indicates that the corresponding variable's coefficient in the model is included in the contrast. It means you are interested in estimating the effect of that variable.

0: This value indicates that the corresponding variable's coefficient in the model is not included in the contrast. It means you are not interested in estimating the effect of that variable.

-1: This value indicates that the corresponding variable's coefficient in the model is included in the contrast, but with an opposite sign. It means you are interested in estimating the negative effect of that variable.

----------------------------------------------------------------
The contrast matrix is typically a matrix with dimensions (number of contrasts) x (number of regression coefficients). Each row of the contrast matrix represents a contrast or comparison you want to test.

For example, let's say you have the following regression coefficients in your model:

Intercept, Age, connectivity, Age_interaction_connectivity
A contrast matric has dimensions of [n_predictors, n_experiments] where each experiment is a contrast

If you want to test the hypothesis that the effect of Age is significant, you can set up a contrast matrix with a row that specifies this contrast (actually an averaging vector):
```
[0,1,0,0]. This is an averaging vector because it sums to 1
```
This contrast will test the coefficient corresponding to the Age variable against zero.


If you want to test the hypothesis that the effect of Age is different from the effect of connectivity, you can set up a contrast matrix with two rows:
```
[0,1,−1,0]. This is a contrast because it sums to 0
```

Thus, if you want to see if any given effect is significant compared to the intercept (average), you can use the following contrast matrix:
```
[1,0,0,0]
[-1,1,0,0]
[-1,0,1,0]
[-1,0,0,1] actually a coding matrix of averaging vectors
```

The first row tests the coefficient for Age against zero, and the second row tests the coefficient for connectivity against zero. The difference between the two coefficients can then be assessed.
_____
You can define any number of contrasts in the contrast matrix to test different hypotheses or comparisons of interest in your regression analysis.

It's important to note that the specific contrasts you choose depend on your research questions and hypotheses. You should carefully consider the comparisons you want to make and design the contrast matrix accordingly.

- Examples:
    - [Two Sample T-Test](https://fsl.fmrib.ox.ac.uk/fsl/fslwiki/GLM#Two-Group_Difference_.28Two-Sample_Unpaired_T-Test.29)
    - [One Sample with Covariate](https://fsl.fmrib.ox.ac.uk/fsl/fslwiki/GLM#Single-Group_Average_with_Additional_Covariate)

In [17]:
contrast_matrix = cal_palm.generate_basic_contrast_matrix(design_matrix)

Here is a basic contrast matrix set up to evaluate the significance of each variable.
Here is an example of what your contrast matrix looks like as a dataframe: 
   DX_BASELINE_Alzheimer  DX_BASELINE_MCI  DX_BASELINE_Normal
0                      1                0                   0
1                      0                1                   0
2                      0                0                   1
Below is the same contrast matrix, but as an array.
Copy it into a cell below and edit it for more control over your analysis.
[
    [1, 0, 0],
    [0, 1, 0],
    [0, 0, 1],
]


In [18]:
# contrast_matrix = [
#     [1, -1, 0],
#     [1, 0]
# ]

In [19]:
contrast_matrix_df = cal_palm.finalize_contrast_matrix(design_matrix=design_matrix, 
                                                    contrast_matrix=contrast_matrix) 
contrast_matrix_df

,DX_BASELINE_Alzheimer,DX_BASELINE_MCI,DX_BASELINE_Normal
0,1,0,0
1,0,1,0
2,0,0,1


# 05 - Generate Files
Standardization during regression is critical. 
- data_transform_method='standardize' will ensure the voxelwise values are standardized
    - if you design matrix has a column called 'Dataset', the standardization will standardize values within each dataset individually, which is as should be done normally.
    - If you call data_transform_method='standardize' without having a 'Dataset' column in your design matrix, the entire collection of images will be standardized. This is potentially dangerous and misleading. Be careful, and consider not standardizing at all, or going back and adding a 'Dataset' column. 

Mask Path
- set mask_path to the path of your local brain mask which matches the resolution of the files you have collected. Typically this is an MNI 152 brain mask. 
    - download one here: https://nilearn.github.io/dev/modules/generated/nilearn.datasets.load_mni152_brain_mask.html

In [20]:
mask_path = '/Users/cu135/Software_Local/calvin_utils_project/circuit_pyper/resources/MNI152_T1_2mm_brain_mask.nii'

Define exchangeability block
- Set to none if you don't know
- If you are running multiple cohorts, set exchangeability block to be the column which has each group in it, with groups being indicated by integers. 

In [21]:
exchangeability_col = None

In [22]:
from calvin_utils.permutation_analysis_utils.voxelwise_regression_prep import RegressionPrep
preparer = RegressionPrep(design_matrix=design_matrix, 
                          contrast_matrix=contrast_matrix_df, 
                          outcome_df=outcome_df, 
                          out_dir=out_dir,
                          voxelwise_variables=voxelwise_variable_list, 
                          voxelwise_interactions=voxelwise_interaction_terms,
                          mask_path=mask_path, 
                          exchangeability_block=None, 
                          data_transform_method=None,
                          weights=None)
dataset_dict, json_path = preparer.run()

Loading & masking: 100%|██████████| 175/175 [00:03<00:00, 46.66it/s]


design_tensor shape: (175, 3, 1)
outcome_data shape: (175, 1, 228483)


# 06 - Run the Voxelwise Regression

In [ ]:
from calvin_utils.permutation_analysis_utils.voxelwise_regression import VoxelwiseRegression
regression = VoxelwiseRegression(json_path, mask_path=mask_path, out_dir=out_dir, regression_type='linear', n_permutations=0)
results = regression.run_all_outputs()

Outcome is not all binary. You should ensure regression_type='linear'. Detected regression_type: linear

Running regression for output 0


Test the voxelwise model
- Run a LOOCV, generating prediction maps

In [ ]:
from calvin_utils.permutation_analysis_utils.voxelwise_regression import VoxelwiseRegression
regression = VoxelwiseRegression(json_path, mask_path=mask_path, out_dir=out_dir+'/loocv_predictions', regression_type='linear', n_permutations=0)
results = regression.run_prediction_loocv()

Predict Another Dataset
- Import another dataset (currently cannot normalized to the other data's distribution, so make sure they are on the same scale)

In [ ]:
from calvin_utils.permutation_analysis_utils.voxelwise_regression_prep import RegressionPrep
preparer = RegressionPrep(design_matrix=design_matrix, 
                          contrast_matrix=contrast_matrix_df, 
                          outcome_df=outcome_df, 
                          out_dir=out_dir,
                          voxelwise_variables=voxelwise_variable_list, 
                          voxelwise_interactions=voxelwise_interaction_terms,
                          mask_path=mask_path, 
                          exchangeability_block=None, 
                          data_transform_method=None,
                          weights=None)
test_dataset_dict, test_json_path = preparer.run()

- Run the fitted voxelwise model to generate prediction maps

In [ ]:
from calvin_utils.permutation_analysis_utils.voxelwise_regression import VoxelwiseRegression
regression = VoxelwiseRegression(test_json_path, mask_path=mask_path, out_dir=out_dir+'/test_predictions', regression_type='linear', n_permutations=0)
results = regression.run_prediction(params_dir=out_dir)

That's all

-Calvin